### PR TITLE
Added Role-based access control integration tests for Spanner Change …

### PR DESCRIPTION
**Pull request by nuggetwheat**
Expanded SpannerChangeStreamIT integration test to include role-based access control tests. The added tests runs with a separate metadata database to ensure this method of operation works properly.

### CODE REVIEW

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/it/ChangeStreamTestPipelineOptions.java [hunk 1]
```diff
@@ -37,8 +37,20 @@ public interface ChangeStreamTestPipelineOptions extends IOTestPipelineOptions,
...
   @Description("Database ID prefix to write to in Spanner")
-  @Default.String("changestream")
+  @Default.String("cs_primary")
   String getDatabaseId();
 
   void setDatabaseId(String value);
+
+  @Description("Role-Based Access Control Database ID prefix to write to in Spanner")
+  @Default.String("cs_rbac")
+  String getRbacDatabaseId();
```
Comment (hmehrotra) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/it/ChangeStreamTestPipelineOptions.java:47]: Do we need a separate database for RBAC? Why can't we use the same DatabaseId? You would still need to create a separate DB client for rbac.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/it/IntegrationTestEnv.java [hunk 2]
```diff
@@ -48,18 +49,44 @@ public class IntegrationTestEnv extends ExternalResource {
...
   private final String host = "https://spanner.googleapis.com";
   private DatabaseAdminClient databaseAdminClient;
   private DatabaseClient databaseClient;
+  private DatabaseClient rbacDatabaseClient;
   private boolean isPostgres;
+  public boolean createRbacDatabase;
+  public boolean useSeparateMetadataDb;
+
+  public IntegrationTestEnv withRbacDatabase() {
+    this.createRbacDatabase = true;
```
Comment (hmehrotra) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/it/IntegrationTestEnv.java:73]: Use 'isRbac' as the parameter name to be consistent with 'isPostgres'?

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/it/IntegrationTestEnv.java [hunk 3]
```diff
@@ -70,154 +97,186 @@ protected void before() throws Throwable {
...
     databaseClient = spanner.getDatabaseClient(DatabaseId.of(projectId, instanceId, databaseId));
+    databases.add(databaseId);
+
+    if (createRbacDatabase) {
+      recreateDatabase(databaseAdminClient, instanceId, rbacDatabaseId, isPostgres);
+      rbacDatabaseClient =
+          spanner.getDatabaseClient(DatabaseId.of(projectId, instanceId, rbacDatabaseId));
+      databases.add(rbacDatabaseId);
+    }
+    if (useSeparateMetadataDb) {
```
Comment (hmehrotra) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/it/IntegrationTestEnv.java:117]: Do we need a separate param for useSeparateMetadataDb? We could just check if (isRbac) here. Right?

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/it/ChangeStreamTestPipelineOptions.java (hunk 1 above)
Comment (nuggetwheat) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/it/ChangeStreamTestPipelineOptions.java]: Done. We don't need a separate DB client for rbac, just need to specify SpannerConfig.databaseRole when in RBAC mode.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/it/IntegrationTestEnv.java (hunk 2 above)
Comment (nuggetwheat) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/it/IntegrationTestEnv.java]: No longer relevant.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/it/IntegrationTestEnv.java (hunk 3 above)
Comment (nuggetwheat) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/it/IntegrationTestEnv.java]: No longer relevant.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/it/SpannerChangeStreamIT.java [hunk 4]
```diff
@@ -93,6 +101,22 @@ public void before() {
...
+  }
+
+  @Test
+  public void testReadSpannerChangeStreamWithAuthorizedRole() {
+    testReadSpannerChangeStreamImpl(pipeline, ENV.getDatabaseRole());
+  }
+
+  @Test
+  public void testReadSpannerChangeStreamWithUnauthorizedRole() {
+    exception.expect(SpannerException.class);
```
Comment (Abacn) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/it/SpannerChangeStreamIT.java:114]: the added tests will be executed on both direct runner precommit and dataflow postcommit:

https://ci-beam.apache.org/view/PostCommit/job/beam_PreCommit_Java_GCP_IO_Direct_Cron

https://ci-beam.apache.org/view/PostCommit/job/beam_PostCommit_Java_DataflowV2

For the latter the exception would not thrown. If remote pipeline fails the PipelineResult is set to FAIL.

Comment (Abacn) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/it/SpannerChangeStreamIT.java:114]: If we only need to validate the error handling on direct runner we can add something like `assumeTrue(options.getRunner() == DirectRunner.class);`

Comment (nuggetwheat) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/it/SpannerChangeStreamIT.java:117]: Thanks for pointing this out! I've added the assumeTrue statement.

**Comment by nuggetwheat** [comment]
Run Java_GCP_IO_Direct PreCommit

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/it/SpannerChangeStreamIT.java [hunk 5]
```diff
@@ -93,6 +103,23 @@ public void before() {
 
   @Test
   public void testReadSpannerChangeStream() {
+    testReadSpannerChangeStreamImpl(pipeline, null);
+  }
+
+  @Test
+  public void testReadSpannerChangeStreamWithAuthorizedRole() {
```
Comment (hmehrotra) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/it/SpannerChangeStreamIT.java:110]: Have we tested that if someone breaks the pattern of not using MetadataConfig constructor, then this test fails?

Comment (nuggetwheat) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/it/SpannerChangeStreamIT.java:110]: Yes. If someone (i.e. Apache Beam code) doesn't use the MetadataConfig constructor to generate the config to access the metadata DB in such a way that breaks access to the metadata DB, this test will catch it.  It is precisely why I modified this test to use a separate metadata database.

**Review by hmehrotra — APPROVED** [review]
LGTM.

**Comment by kennknowles** [comment]
R: @johnjcasey

**Comment by kennknowles** [comment]
John, can you review or ping the best person? (looks pretty easy but I want to respect boundaries of my expertise)

**Review by Abacn — APPROVED** [review]
tests passed. this lgtm, thanks

**Pull request merged**